pmd-java/src/main/java/net/sourceforge/pmd/lang/java/JavaProcessingStage.java

```diff
@@ -0,0 +1,134 @@
+
+    /**
+     * Qualified name resolution.
+     */
+    QNAME_RESOLUTION("Qualified name resolution") {
+        @Override
+        public void processAST(RootNode rootNode, AstAnalysisConfiguration configuration) {
+            new QualifiedNameResolver().initializeWith(configuration.getTypeResolutionClassLoader(), (ASTCompilationUnit) rootNode);
+        }
+    },
+
+    /**
+     * Symbol table analysis.
+     */
+    SYMBOL_RESOLUTION("Symbol table") {
+        @Override
+        public void processAST(RootNode rootNode, AstAnalysisConfiguration configuration) {
+            new SymbolFacade().initializeWith(configuration.getTypeResolutionClassLoader(), (ASTCompilationUnit) rootNode);
+        }
+    },
+
+    /**
+     * Type resolution, depends on QName resolution.
+     */
+    TYPE_RESOLUTION("Type resolution", QNAME_RESOLUTION) {
+        @Override
+        public void processAST(RootNode rootNode, AstAnalysisConfiguration configuration) {
+            new TypeResolutionFacade().initializeWith(configuration.getTypeResolutionClassLoader(), (ASTCompilationUnit) rootNode);
+        }
+    },
+
+    /**
+     * Data flow analysis.
+     */
+    DFA("Data flow analysis") {
+        @Override
+        public void processAST(RootNode rootNode, AstAnalysisConfiguration configuration) {
+            new DataFlowFacade().initializeWith(new JavaDataFlowHandler(), (ASTCompilationUnit) rootNode);
+        }
+
+
+        @Override
+        public boolean dependsOnImpl(Rule rule) {
+            return rule.isDfa();
+        }
+    };
+
+
+    private final String displayName;
+    private final List<JavaProcessingStage> dependencies;
+
+
+    JavaProcessingStage(String displayName, JavaProcessingStage... dependencies) {
+        this.displayName = displayName;
+        this.dependencies = Collections.unmodifiableList(Arrays.asList(dependencies));
+    }
+
+
+    @Override
+    public List<JavaProcessingStage> getDependencies() {
+        return dependencies;
+    }
+
+
+    @Override
+    public String getDisplayName() {
+        return displayName;
+    }
+
+
+    /**
+     * Returns true if the given Java rule depends on this stage.
+     *
+     * <p>{@link AbstractJavaRule#dependsOn(AstProcessingStage)}
+     * delegates to this implementation after a validity check.
+     * Dispatching on the rule language & forwarding to the
+     * processing stage implementation allows specializing XPath
+     * rules differently from rules written in Java, while keeping
+     * dependency specification inside the processing stage declaration,
+     * for readability. By design, Java XPath rules ignore this
+     * method, it is only relevant to {@link AbstractJavaRule}.
+     *
+     * @param rule Rule to check
+     *
+     * @return True if the given rule depends on this stage
+     *
+     * @throws IllegalArgumentException if this rule is not a rule for Java.
+     */
+    @Experimental
+    public final boolean ruleDependsOnThisStage(Rule rule) {
```

Comment:
Yeah but it's the only way I know to implement double-dispatch in java... It's just like a visitor pattern tbh, the data flow is also unintuitive there.

Reply:
wouldn't it be best if we have the rule directly inform a:

```
Set<AstProcessingStage<T>> getRequestedProcessingStages() 
```

`T` would have to be bound at the base abstract rule per language.

This way, the rule knows the stages it needs, without the stages to need to know the rules; and breaking the tight coupling. This should be fine, as rules are dynamic (ie: can be added through plugins), but processing stages cannot, as they are part of the enum.

I'm unsure we actually need double dispatching here.

Reply:
Double dispatch here is mostly a way to group the implementation of "how does a rule declare a dependency on a stage" with the implementation of the stage, ie in the enum instead of the rule. I thought the code would be clearer that way but ultimately you're right, we only need to dispatch on the rule's type. Ignoring the enum's `ruleDependsOnThisStage` for XPath rules also seems sloppy. If you'd like to do the change yourself please go ahead, otherwise let me know

I'm not sure adding a type parameter is worth it though. The processing stages will nearly exclusively be handled in pmd-core, where no information on the value of the type argument is available. The extra type safety would disappear, and force us to work around it with wildcards and stuff. So we couldn't avoid runtime checks anyway.